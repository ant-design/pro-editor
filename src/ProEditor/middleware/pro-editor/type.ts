import { StateCreator } from 'zustand/vanilla';
import { ConfigPublicAction } from '../../store/slices/config';
import { TakeTwo, Write } from '../types/utils';

export interface ProEditorSetStateAction {
  type: unknown;

  /**
   * 如果记录，那么历史记录的名字是什么
   */
  name?: string;
  /**
   * 是否将操作记录到历史记录中
   */
  recordHistory?: boolean;
}

/**
 * 用于给注入方法添加额外的 proEditor 实例对象
 */
export interface ProEditorMiddlewareInjectMethod<Sa extends unknown[], Sr, T, A> {
  setState<A1 extends string | ProEditorSetStateAction>(...a: [...a: TakeTwo<Sa>, action?: A1]): Sr;

  proEditor: {
    undo?: () => void;
    redo?: () => void;
    // clearStorage: () => void;
    getOptions: () => Partial<ProEditorOptions<T, A>>;
  };
}
/**
 * 提供给用户的配置项
 */
export interface ProEditorOptions<S, EditorSaveState = S> {
  /** Name of the storage (must be unique) */
  name: string;
  /**
   * Filter the persisted value.
   *
   * @params state The state's value
   */
  partialize?: (state: S) => EditorSaveState;
}

// 为  mutator 注入 'pro-editor' 类型，以支持第三个配置参数

type WithProEditor<S, A> = S extends {
  getState: () => infer T;
  setState: (...a: infer Sa) => infer Sr;
}
  ? Write<S, ProEditorMiddlewareInjectMethod<Sa, Sr, T, A>>
  : never;

declare module 'zustand/vanilla' {
  interface StoreMutators<S, A> {
    ['pro-editor']: WithProEditor<S, A>;
  }
}

// 内部方法，用于保证中间件内部的类型定义安全

export type ProEditorImpl = <T>(
  storeInitializer: StateCreator<T, [['pro-editor', unknown]], []>,
  options: ProEditorOptions<T, T>,
) => StateCreator<T, [['pro-editor', unknown]], []>;

export interface InjectInternalProEditor {
  proEditor: {
    __INTERNAL_SET_CONFIG__NOT_USE_IT: ConfigPublicAction['setConfig'];
  };
}
