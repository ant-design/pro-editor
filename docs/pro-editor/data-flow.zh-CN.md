---
title: ProEditor 数据流架构
group: 基础框架
---

# 适配多人协同模式的数据流架构

在多人协同中，我们根据数据的用途可以划分出两大类数据模型：

1. **文件数据**：唯一性数据实例（ N 个人看到的都是同一份文件数据）
2. **界面视图**：每个人默认看到自己的视图，但当用户选择跟随其他人时，视图完全 follow 对方的视图）

## ProEditor 容器数据流架构

ProEditor 的数据流包含两类：

- 容器类，有且仅有一个： `ProEditorStore` ;
- 组件类，视根据未来的扩展情况进行扩张，目前有 `ProTableStore`、`ProLayoutStore`，后续可能为会扩展 `ProFromStore`、`ProChartStore` 等；

容器类 Store 只存放 Editor 相关状态，例如：

- 开发者 / 设计师模式；
- 画布 / 代码的切换；
- 当前画布交互的激活信息；
- 当前加载的 ComponentAsset 实例；
- ...

组件类 Store 会包含具体的组件配置，例如：

- 组件配置信息；
- 组件面板切换的 tabs；
- 预览器折叠状态；
- ...

这样拆分的目的就是将关注点进行分离，进而方便未来的扩展。

## 数据流流转模式

数据流连接方式，分别从容器层和组件层进行说明。

### 容器层

容器层核心的作用是调度组件层数据与外部数据源的通信。

由于 ProEditor 本身是一个组件，因此它必须同时支持受控模式与非受控模式，这对于数据流的架构就带来了很大的挑战。

由于编辑器的信息处理较为复杂，非受控模式下，数据全部由组件自行处理是没有问题。但是一旦到了受控模式，内部状态和外部状态应该如何同步才能保证数据的一致性，并让组件消费者使用上做到较为灵活，便是一件很头疼的问题。

像 `ProTableStore` 里的 columns 数据有 20+ 的 dispatch 方法， 如果这些方法如果直接暴露给组件外部消费的开发者，外部开发者怕是要炸。

在之前的踩坑过程中，如果使用 useEffect 监听外部变化 setState 进来，内部 state 变化时 onChange 出去，非常容易造成重复渲染的死循环，这对于开发者的心智负担极重。

下面的代码在较为简单的场景下可以适用，但是在我们有 10+ 操作方法，并且需要使用 reducer 的情况下，数据流的处理就会变成一团乱麻。

```tsx | pure
import { useEffect, useState } from 'react';

const useStore = (value, onChange) => {
  const [innerValue, setValue] = useState('');

  useEffect(() => {
    setValue(value);
  }, [value]);

  const innerSetState = (value) => {
    setValue(value);

    onChange(value);
  };
};
```

因此，从最终的开发上，采用以内部数据源作为 SoT ，将外部的数据源根据情况同步进内部数据源，实现状态管理。

具体的开发策略为：

1. 首先和正常开发一样，在 Wrapper 创建一个顶层数据源，下发给子组件；
2. 将组件本身作为一个非受控组件进行开发（App），不考虑任何外部受控值；
3. 使用 `StoreUpdater` 组件作为**胶水层**隔离内部应用（`App`）与 外部数据源（`config`、`onConfigChange`、`defaultConfig`等）；
4. 根据情况在 `store` 中确定需要数据交互的方法、对象，在`internalUpdate`方法中加入`OnChange`，用`syncOutSource`方法同步外部数据源到内部；

```tsx | pure
export const Container: FC = (props) => {
  const {
    config,
    onConfigChange,
    defaultConfig,

    ...res
  } = props;

  return (
    <Wrapper>
      <App {...res} />
      <StoreUpdater config={config} defaultConfig={defaultConfig} onConfigChange={onConfigChange} />
    </Wrapper>
  );
};
```
