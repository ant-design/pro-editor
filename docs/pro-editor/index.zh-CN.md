---
title: ProBuilder 配置容器
atomId: ProBuilder
group: 组件
nav:
  title: 装配器
  order: 3
---

# ProBuilder 装配器

使用 `ProBuilder` 可以快速构建出组件的配置界面。

## 代码演示

<code src='./demos/buttonAssets.tsx' compact="true" title="Button 示例"></code>

<code src='./demos/defaultAssets.tsx' title="传入 `ComponentAssets`"></code>

<code src='./demos/empty.tsx' title="无 `ComponentAssets`"></code>

<code src="./demos/controlledPresence.tsx" title="视图状态受控" description="将 `editorAwareness` 设为受控状态，可以控制当前的视窗状态。"></code>

## API

<API id="ProEditor"></API>
