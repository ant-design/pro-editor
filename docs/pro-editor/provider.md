---
title: ProEditorProvider
group: 组件
---

# ProEditorProvider

如果：

- 希望在不是 ProEditor 组件子元素的组件中使用 `useProEditor`、useConfig、`useViewport` 等 ProEditor 钩子；
- 在页面上有多个 ProEditor 实例；
- 需要在不是 ProEditor 组件子元素的组件中访问内部状态，或使用 ProEditor 所提供的数据方法。

那么你就需要用 `ProEditorProvider` 包裹相应的应用组件。
