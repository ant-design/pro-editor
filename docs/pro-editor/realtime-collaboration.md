---
title: 实时协同模块
group: 容器能力
---

# 实时协同模块

实现了多人实时协同的基础功能的模块。

- 数据同步（ yjs-zustand 中间件）
- 协同感知（ Avatar、Cursor 组件）

## Demo 演示

<code src="./demos/realtimeCollaboration/demo.tsx"></code>

## 数据同步： Yjs 中间件

复用了 [zustand-middleware-yjs](https://github.com/joebobmiles/zustand-middleware-yjs) 中间件

后面只要为 ProEditor 容器设计合适的数据同步分层结构，ProEditor 容器就可以具有多人协作的能力。

## 协同感知

zustand-middleware-yjs 并不提供 yjs 中的 Awareness 能力。因此比如在多人协同中比较常用的用户状态、指针显示等功能是自行实现的。

## 历史记录（TODO）

历史记录在基于 YJS 实现多人协同后，是属于天生自带的功能。但这个历史记录只是大量的数据，这些数据该如何合适地对用户展示，还需要设计合理的交互方案。

## 离线缓存

离线缓存 是 Local First 理念的核心之一
