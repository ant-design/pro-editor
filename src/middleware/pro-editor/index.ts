// import {  } from '@/ProEditor/store';
import { StateCreator, StoreMutatorIdentifier } from 'zustand/vanilla';
import { ProEditorImpl, ProEditorSetStateAction } from './type';

/**
 * 提供给用户的配置项
 */
export interface ProEditorOptions<S, EditorSaveState = S> {
  /** Name of the storage (must be unique) */
  name: string;
  /**
   * Filter the persisted value.
   *
   * @params state The state's value
   */
  partialize?: (state: S) => EditorSaveState;
}

const middleware: ProEditorImpl = (storeInitializer, options) => (set, get, api) => {
  const partialize = options.partialize ?? ((s) => s);
  /**
   * 记录历史
   * @param action
   */
  const updateInProEditor = (action: ProEditorSetStateAction) => {
    const nextConfig = partialize(get());

    get().proEditor.__INTERNAL_SET_CONFIG__NOT_USE_IT(nextConfig, {
      trigger: 'proEditorMiddleware',
      ...action,
    });
  };
  /*
   * Capture the initial state so that we can initialize the pro editor store to the
   * same values as the initial values of the Zustand store.
   */
  const store = storeInitializer(
    /*
     * Create a new set function that defers to the original and then passes
     * the new state to patchSharedType.
     */
    (partial, replace, action) => {
      set(partial, replace, action);
      updateInProEditor((action as any) || {});
    },
    get,
    {
      ...api,
      // Create a new setState function as we did with set.
      setState: (partial, replace, action) => {
        api.setState(partial, replace, action);

        updateInProEditor((action as any) || {});
      },
    },
  );

  // Return the initial state to create or the next middleware.
  return {
    ...store,
    proEditor: { options: { ...options, partialize } },
  };
};

export type ProEditorMiddleware = <
  T,
  Mps extends [StoreMutatorIdentifier, unknown][] = [],
  Mcs extends [StoreMutatorIdentifier, unknown][] = [],
  U = T,
>(
  initializer: StateCreator<T, [...Mps, ['pro-editor', unknown]], Mcs>,
  options: ProEditorOptions<T, U>,
) => StateCreator<T, Mps, [['pro-editor', U], ...Mcs]>;

export const proEditorMiddleware = middleware as unknown as ProEditorMiddleware;
