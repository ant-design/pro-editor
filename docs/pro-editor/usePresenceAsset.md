---
title: usePresenceAsset
group: Hooks
order: 2
---

该方法可以获取编辑器中的资产的感知视图状态

```ts
import { usePresenceAsset } from '@ant-design/pro-editor';

const getPresenceAsset = () => {
  const [presence, setPresence] = usePresenceAsset();

  useEffect(() => {
    console.log(setPresence);
  }, [presence]);

  return null;
};
```

:::warning

注：该 hook 只能在被 `ProEditorProvider` 包裹的组件中使用，或者在 `<ProEditor />` 组件的子组件中使用。

:::
