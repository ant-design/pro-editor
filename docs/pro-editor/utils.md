---
title: C2D2C 辅助函数
group:
  title: 工具类库
  order: 99
---

# generateJSXCode

该方法用于将组件属性配置拼接生成 jsx 代码。

```jsx | pure
import { generateJSXCode } from '@ant-design/pro-editor';

const config = {
  children: '默认按钮',
  target: '_blank',
  type: 'default',
  htmlType: 'button',
  danger: false,
  icon: '',
  size: 'middle',
  loading: true,
  disabled: false,
  ghost: false,
  block: false,
};

const code = generateJSXCode('Button', config);
```

结果为:

```jsx | pure
<Button
  target="_blank"
  type="default"
  htmlType="button"
  danger={false}
  size="middle"
  loading
  disabled={false}
  ghost={false}
  block={false}
>
  默认按钮
</Button>
```
