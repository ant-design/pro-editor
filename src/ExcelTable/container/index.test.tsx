import { ExcelTable } from '@ant-design/pro-editor';
import { render } from '@testing-library/react';

// const tableData = [
//   ['序号', '标题', '状态', '标签', '创建时间', '操作'],
//   ['1', '你好1', '成功', 'Tag1', '昨天', '修改'],
//   ['2', '2你好', '失', 'Tag2', '今天', '修改'],
//   ['3', '你好3', '失败', 'Tag4', '明天', '修改'],
//   ['4', '4你好', '成功', 'Tag5', '大后天', '修改'],
// ];

describe('ExcelTable', () => {
  it('可正常渲染', () => {
    const { container } = render(<ExcelTable />);

    // 由于 handsontable 每次会创建一个随机 id，这会导致快照不准确，所以把相应的id号去掉
    const ctn = container.getElementsByClassName('htRowHeaders').item(0);
    // 去掉随机的 id 号
    ctn.id = 'id-';

    expect(container).toMatchSnapshot();
  });

  // it('可渲染默认值', async () => {
  //   const { findAllByText } = render(<ExcelTable defaultData={tableData} />);
  //
  //   const btns = await findAllByText('修改');
  //   console.log(btns);
  //
  //   expect(btns.length).toEqual(4);
  // });

  //   it('受控模式可正常受控', async () => {
  //     const Component = () => {
  //       const [data, setData] = useState<TableData>();

  //       return (
  //         <div>
  //           <Button
  //             onClick={() => {
  //               setData(tableData);
  //             }}
  //             data-testid={'btn'}
  //           >
  //             插入
  //           </Button>
  //           <Divider />
  //           <ExcelTable data={data} onDataChange={setData} />
  //         </div>
  //       );
  //     };
  //     const { findByTestId, container } = render(<Component />);

  //     const btn = await findByTestId('btn');

  //     await act(async () => {
  //       await userEvent.click(btn);
  //     });

  //     // 由于 handsontable 每次会创建一个随机 id，这会导致快照不准确，所以把相应的id号去掉
  //     const ctn = container.getElementsByClassName('htRowHeaders').item(0);
  //     // 去掉随机的 id 号
  //     ctn.id = 'id-';

  //     expect(container).toMatchSnapshot();
  //   });
});
