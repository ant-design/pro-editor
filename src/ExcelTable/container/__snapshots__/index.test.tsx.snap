// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExcelTable > 可正常渲染 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  width: 100%;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 8px;
}

.emotion-3 {
  width: 90px;
  background: rgba(0, 0, 0, 0.02);
  border-color: transparent;
  box-shadow: none;
}

.emotion-3 .studio-input-number-group-addon {
  padding: 0 4px;
  color: rgba(0, 0, 0, 0.65);
  background: rgba(0, 0, 0, 0.02);
  border-color: transparent;
}

.emotion-8 .handsontable th:first-child,
.emotion-8 .handsontable th:nth-child(2),
.emotion-8 .handsontable td:first-of-type {
  border: none;
}

.emotion-8 .handsontable {
  color: rgba(0, 0, 0, 0.65);
  font-size: 12px;
}

.emotion-8 .handsontable .htCore {
  overflow: scroll;
  border-radius: 4px;
}

.emotion-8 .handsontable .handsontable th:last-child {
  border-right-color: rgba(0, 0, 0, 0.04);
}

.emotion-8 .handsontable th,
.emotion-8 .handsontable td {
  background: none;
  border-top: none!important;
  border-right-color: rgba(0, 0, 0, 0.04);
  border-bottom: none;
  border-left-color: rgba(0, 0, 0, 0.04);
}

.emotion-8 .handsontable th,
.emotion-8 .handsontable tr:first-child {
  color: rgba(0, 0, 0, 0.88);
  background: rgba(0, 0, 0, 0.04);
  border-top: rgba(0, 0, 0, 0.06);
}

.emotion-8 .handsontable tr:nth-child(2n + 1) {
  background: rgba(0, 0, 0, 0.04);
}

<div>
  <div
    class="layoutkit-flexbox emotion-0 studio-excel-table"
  >
    <div
      class="layoutkit-flexbox emotion-1"
    >
      <div
        class="layoutkit-flexbox emotion-2"
      >
        <div
          class="studio-input-number-group-wrapper studio-input-number-group-wrapper-sm studio-excel-table-input emotion-3"
        >
          <div
            class="studio-input-number-wrapper studio-input-number-group"
          >
            <div
              class="studio-input-number studio-input-number-sm studio-excel-table-input emotion-3"
            >
              <div
                class="studio-input-number-handler-wrap"
              >
                <span
                  aria-disabled="false"
                  aria-label="Increase Value"
                  class="studio-input-number-handler studio-input-number-handler-up"
                  role="button"
                  unselectable="on"
                >
                  <span
                    aria-label="up"
                    class="anticon anticon-up studio-input-number-handler-up-inner"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="up"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  aria-disabled="false"
                  aria-label="Decrease Value"
                  class="studio-input-number-handler studio-input-number-handler-down"
                  role="button"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down studio-input-number-handler-down-inner"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
              <div
                class="studio-input-number-input-wrap"
              >
                <input
                  aria-valuemin="1"
                  aria-valuenow="5"
                  autocomplete="off"
                  class="studio-input-number-input"
                  role="spinbutton"
                  step="1"
                  value="5"
                />
              </div>
            </div>
            <div
              class="studio-input-number-group-addon"
            >
              列
            </div>
          </div>
        </div>
        <div
          class="studio-input-number-group-wrapper studio-input-number-group-wrapper-sm studio-excel-table-input emotion-3"
        >
          <div
            class="studio-input-number-wrapper studio-input-number-group"
          >
            <div
              class="studio-input-number studio-input-number-sm studio-excel-table-input emotion-3"
            >
              <div
                class="studio-input-number-handler-wrap"
              >
                <span
                  aria-disabled="false"
                  aria-label="Increase Value"
                  class="studio-input-number-handler studio-input-number-handler-up"
                  role="button"
                  unselectable="on"
                >
                  <span
                    aria-label="up"
                    class="anticon anticon-up studio-input-number-handler-up-inner"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="up"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  aria-disabled="false"
                  aria-label="Decrease Value"
                  class="studio-input-number-handler studio-input-number-handler-down"
                  role="button"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down studio-input-number-handler-down-inner"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
              <div
                class="studio-input-number-input-wrap"
              >
                <input
                  aria-valuemin="1"
                  aria-valuenow="5"
                  autocomplete="off"
                  class="studio-input-number-input"
                  role="spinbutton"
                  step="1"
                  value="5"
                />
              </div>
            </div>
            <div
              class="studio-input-number-group-addon"
            >
              行
            </div>
          </div>
        </div>
      </div>
      <div>
        <button
          class="studio-btn studio-btn-default studio-btn-sm studio-excel-table-btn"
          type="button"
        >
          <span>
            重 置
          </span>
        </button>
      </div>
    </div>
    <div
      class="layoutkit-flexbox emotion-0"
      style="margin-top: 12px;"
    >
      <div
        class="studio-excel-table-table emotion-8"
      >
        <div
          class="handsontable htRowHeaders"
          data-originalstyle="height: auto; overflow: hidden; width: 100%;"
          dir=""
          id="id-"
          style="height: auto; overflow: hidden; width: 100%;"
        >
          <div
            class="ht_master handsontable"
            dir="ltr"
          >
            <div
              class="wtHolder"
              style="position: relative; height: auto; width: 0px;"
            >
              <div
                class="wtHider"
                style="width: 300px; height: 139px;"
              >
                <div
                  class="wtSpreader"
                  style="position: relative; top: 0px; left: 0px;"
                >
                  <table
                    class="htCore"
                  >
                    <colgroup>
                      <col
                        class="rowHeader"
                        style="width: 50px;"
                      />
                      <col
                        style="width: 50px;"
                      />
                    </colgroup>
                    <thead />
                    <tbody>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              列名称
                            </span>
                          </div>
                        </th>
                        <td
                          class=""
                        >
                          
                        </td>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              1
                            </span>
                          </div>
                        </th>
                        <td
                          class=""
                        >
                          
                        </td>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              2
                            </span>
                          </div>
                        </th>
                        <td
                          class=""
                        >
                          
                        </td>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              3
                            </span>
                          </div>
                        </th>
                        <td
                          class=""
                        >
                          
                        </td>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              4
                            </span>
                          </div>
                        </th>
                        <td
                          class=""
                        >
                          
                        </td>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              5
                            </span>
                          </div>
                        </th>
                        <td
                          class=""
                        >
                          
                        </td>
                      </tr>
                    </tbody>
                  </table>
                  <div
                    class="htBorders"
                  >
                    <div
                      style="position: absolute; top: 0px; left: 0px;"
                    >
                      <div
                        class="wtBorder current"
                        style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                      />
                      <div
                        class="wtBorder current"
                        style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                      />
                      <div
                        class="wtBorder current"
                        style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                      />
                      <div
                        class="wtBorder current"
                        style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                      />
                      <div
                        class="wtBorder current corner"
                        style="background-color: rgb(75, 137, 255); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                      />
                    </div>
                    <div
                      style="position: absolute; top: 0px; left: 0px;"
                    >
                      <div
                        class="wtBorder fill"
                        style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                      />
                      <div
                        class="wtBorder fill"
                        style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                      />
                      <div
                        class="wtBorder fill"
                        style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                      />
                      <div
                        class="wtBorder fill"
                        style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                      />
                      <div
                        class="wtBorder fill corner"
                        style="background-color: rgb(255, 0, 0); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ht_clone_top handsontable"
            dir="ltr"
            style="position: absolute; top: 0px; overflow: visible; left: 0px;"
          >
            <div
              class="wtHolder"
              style="position: relative;"
            >
              <div
                class="wtHider"
              >
                <div
                  class="wtSpreader"
                  style="position: relative;"
                >
                  <table
                    class="htCore"
                  >
                    <colgroup />
                    <thead />
                    <tbody />
                  </table>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ht_clone_bottom handsontable"
            dir="ltr"
            style="position: absolute; top: 0px; overflow: visible; left: 0px;"
          >
            <div
              class="wtHolder"
              style="position: relative;"
            >
              <div
                class="wtHider"
              >
                <div
                  class="wtSpreader"
                  style="position: relative;"
                >
                  <table
                    class="htCore"
                  >
                    <colgroup />
                    <thead />
                    <tbody />
                  </table>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ht_clone_inline_start ht_clone_left handsontable"
            dir="ltr"
            style="position: absolute; top: 0px; overflow: visible; left: 0px; height: 0px; width: 0px;"
          >
            <div
              class="wtHolder"
              style="position: relative; height: 0px; width: 3px;"
            >
              <div
                class="wtHider"
                style="height: 139px;"
              >
                <div
                  class="wtSpreader"
                  style="position: relative; top: 0px;"
                >
                  <table
                    class="htCore"
                  >
                    <colgroup>
                      <col
                        class="rowHeader"
                        style="width: 50px;"
                      />
                    </colgroup>
                    <thead />
                    <tbody>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              列名称
                            </span>
                          </div>
                        </th>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              1
                            </span>
                          </div>
                        </th>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              2
                            </span>
                          </div>
                        </th>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              3
                            </span>
                          </div>
                        </th>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              4
                            </span>
                          </div>
                        </th>
                      </tr>
                      <tr>
                        <th
                          class=""
                        >
                          <div
                            class="relative"
                          >
                            <span
                              class="rowHeader"
                            >
                              5
                            </span>
                          </div>
                        </th>
                      </tr>
                    </tbody>
                  </table>
                  <div
                    class="htBorders"
                  >
                    <div
                      style="position: absolute; top: 0px; left: 0px;"
                    >
                      <div
                        class="wtBorder current"
                        style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                      />
                      <div
                        class="wtBorder current"
                        style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                      />
                      <div
                        class="wtBorder current"
                        style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                      />
                      <div
                        class="wtBorder current"
                        style="background-color: rgb(75, 137, 255); height: 2px; width: 2px; display: none;"
                      />
                      <div
                        class="wtBorder current corner"
                        style="background-color: rgb(75, 137, 255); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                      />
                    </div>
                    <div
                      style="position: absolute; top: 0px; left: 0px;"
                    >
                      <div
                        class="wtBorder fill"
                        style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                      />
                      <div
                        class="wtBorder fill"
                        style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                      />
                      <div
                        class="wtBorder fill"
                        style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                      />
                      <div
                        class="wtBorder fill"
                        style="background-color: rgb(255, 0, 0); height: 1px; width: 1px; display: none;"
                      />
                      <div
                        class="wtBorder fill corner"
                        style="background-color: rgb(255, 0, 0); height: 6px; width: 6px; border: 1px solid #fff; display: none;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ht_clone_top_inline_start_corner ht_clone_top_left_corner handsontable"
            dir="ltr"
            style="position: absolute; top: 0px; overflow: visible; left: 0px; height: 0px; width: 0px;"
          >
            <div
              class="wtHolder"
              style="position: relative;"
            >
              <div
                class="wtHider"
              >
                <div
                  class="wtSpreader"
                  style="position: relative;"
                >
                  <table
                    class="htCore"
                  >
                    <colgroup />
                    <thead />
                    <tbody />
                  </table>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ht_clone_bottom_inline_start_corner ht_clone_bottom_left_corner handsontable"
            dir="ltr"
            style="position: absolute; overflow: visible; left: 0px; height: 0px; width: 0px;"
          >
            <div
              class="wtHolder"
              style="position: relative;"
            >
              <div
                class="wtHider"
              >
                <div
                  class="wtSpreader"
                  style="position: relative;"
                >
                  <table
                    class="htCore"
                  >
                    <colgroup />
                    <thead />
                    <tbody />
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
